doc/cprover-manual/requires-and-ensures-clauses.md

```diff
@@ -0,0 +1,141 @@
+# Requires \& Ensures Clauses
+
+
+### Definition
+
+```c
+__CPROVER_requires(bool cond)
+```
+
+The _requires_ clause specify the preconditions of a function, i.e., the minimal
+requirements to properly execute an operation. Developers may see the _requires_
+clause as all the caller's obligations when invoking this fuction.
+
+```c
+__CPROVER_ensures(bool cond)
+```
+
+The _ensures_ clause specify the postconditions of a function, i.e., all the
+properties over arguments or global variables that a function guarantees at the
+an of the operation.
+
+Also, notice that one may call
+multiple instances of `__CPROVER_requires` and `__CPROVER_ensures` clauses, which will be combined together in a single conjunction.
+
+
+### Parameters
+
+The _requires_ clause accepts a Boolean expression over the arguments of a function and/or global variables, including CBMC primitive functions (e.g., [Memory Predicates]()). Similarly, _ensures_ clauses also accept Boolean expressions and CBMC primitives, but also [History Variables]().
+
+**Important.** Developers may call functions inside _requires_ and _ensures_
+clauses to better write larger specifications (e.g., predicates). However, at
+this point CBMC does not enforce such functions to be without
+side effects (i.e., do not modify any local or global state). This will be supported in future versions.
+
+
+### Semantics
+
+The semantics of a _requires_ clause *r* of a given function *f* can be
+understood in two contexts: enforcement and replacement. 
+To illustrate these two perspectives, consider the following implementation of the `sum` function.
+
+```c
+int sum(uint32_t* a, uint32_t* b, uint32_t* out)
+/* Precondition */
+__CPROVER_requires(__CPROVER_is_fresh(a, sizeof(*a)))
+__CPROVER_requires(__CPROVER_is_fresh(b, sizeof(*b)))
+__CPROVER_requires(__CPROVER_is_fresh(out, sizeof(*out)))
+/* Postconditions */
+__CPROVER_ensures(__CPROVER_return_value == SUCCESS || __CPROVER_return_value == FAILURE)
+__CPROVER_ensures((__CPROVER_return_value == SUCCESS) ==> (*out == (*a + *b)))
+{
+    const uint64_t result = ((uint64_t) *a) + ((uint64_t) *b);
+    if (result > UINT32_MAX) return FAILURE;
+    *out = (uint32_t) result;
+    return SUCCESS;
+}
+```
+
+#### Enforcement
+
+In order to determine whether _requires_ and _ensures_ clauses are a sound
+characterization of the behavior of a function *f*, CBMC will try to check them as follows:
+
+1. Considers all arguments and global variables as non-determinisc values;
+2. Assumes all preconditions specified in the `__CPROVER_requires` clauses;
+4. Calls the implementation of function *f*;
+5. Asserts all postconditions described in the `__CPROVER_ensures` clauses.
+
+In our example, `sum` function will be instrumented as follows
+
+```c
+/* Function Contract Enforcement */
+int add_overflow(uint32_t a, uint32_t b, uint32_t* out)
+{
+  __CPROVER_assume(out != NULL);
+  
+  int return_value_add_overflow = __CPROVER_contracts_original_add_overflow(a, b, out);
```

Comment:
Why not name the enforcement function as `__CPROVER_contracts_enforce_add_overflow` and have this line as:
```c
  int return_value_add_overflow = add_overflow(a, b, out);
```
i.e. not rename the original function, but have a new name for the enforcing function.

(We have an open issue regarding this.)

Reply:
Yes, I agree. However, shouldn't we update the documentation only once we implement this change?

Reply:
You are right. Let's change this after we implement the fix.